### PR TITLE
Detect board in firmware_flasher patch

### PR DESCRIPTION
**Pull request by haslinghuis**
Patch for firmware flashing detect board feature introduced in #2485

Fixes: #2526
Fixes: #2498 

After much testing this patch provides the following changes (thanks for all feedback):

- Removes board verification on tab load and instead let the user check with a button reducing the connection load to avoid connection issues.
- Slapped a timer on the button to prevent spamming as requested by @Asizon
- Button is much smaller now and next to board selection as requested by @McGivergim
- Only present a dialog (when using MSP capabilities) in `msp32.js` when using online firmware as requested by @mikeller, @hydra and @Asizon 
- Like to keep the detection on port change.
- Does not remember non available board.
- Refreshes tab when trying to flash while no board is connected.

Screenshot from 2021-07-01 13-27-24

### CODE REVIEW

**Comment by blckmn** [comment]
AUTOMERGE: (FAIL)
 - github identifies PR as mergeable -> FAIL
 - assigned to a milestone -> PASS
 - cooling off period lapsed -> FAIL
 - commit count less or equal to three -> PASS
 - `Don't merge` label NOT found -> PASS
 - at least one `RN:` label found -> PASS
 - `Tested` label found -> FAIL
 - assigned to an approver -> PASS
 - approver count at least three -> FAIL

File: src/js/serial.js
Comment (McGiverGim) [inline src/js/serial.js:22]: I don't know if the `new Error().stack` will be useful for something. What is the objective?

**Review by McGiverGim — CHANGES_REQUESTED** [review]
I think this approach is far better (but less fancy) than the automatic version. 

I'm not too sure about the "big" button. I think is better to "attach" it in some way to the combo of FCs, but I don't know exactly how. 
Someone has some idea? Maybe a little button with some kind of icon next to the FCs combo, using the space between button an text (maybe given here a little more space)?

File: src/js/serial.js
Comment (haslinghuis) [reply src/js/serial.js:22]: To see where the connection was made from (but is only useful while debugging) - due to timeouts I could not step over the code fast enough, but I will remove it after getting more test results.

**Comment by blckmn** [comment]
AUTOMERGE: (FAIL)
 - github identifies PR as mergeable -> FAIL
 - assigned to a milestone -> PASS
 - cooling off period lapsed -> FAIL
 - commit count less or equal to three -> PASS
 - `Don't merge` label NOT found -> PASS
 - at least one `RN:` label found -> PASS
 - `Tested` label found -> PASS
 - assigned to an approver -> PASS
 - approver count at least three -> FAIL

**Comment by blckmn** [comment]
AUTOMERGE: (FAIL)
 - github identifies PR as mergeable -> FAIL
 - assigned to a milestone -> PASS
 - cooling off period lapsed -> PASS
 - commit count less or equal to three -> PASS
 - `Don't merge` label NOT found -> PASS
 - at least one `RN:` label found -> PASS
 - `Tested` label found -> PASS
 - assigned to an approver -> PASS
 - approver count at least three -> FAIL

**Comment by haslinghuis** [comment]
@McGiverGim can you please rerun failed jobs. I don't have an Azure account

**Review by haslinghuis — COMMENTED** [review]
The button is now dynamic

**Comment by haslinghuis** [comment]
Disabled connection `if reboot_mode === 0` to prevent `serialPortOpenFail` in case user has chosen Abort.

File: src/js/tabs/firmware_flasher.js
Comment (limonspb) [inline src/js/tabs/firmware_flasher.js:836]: wondering if "DFU" stays as "DFU" in all other languages for the element value?

**Review by Asizon — CHANGES_REQUESTED** [review]
@haslinghuis it works so good now, Can we prevent the target from loading 2 times if the button is pressed 2 times quickly? Can you reproduce this issue?

**Comment by haslinghuis** [comment]
@Asizon hiding the button in verifyBoard does almost (95%) do the trick. Could instead slap a timer on the button for 1000ms or adding another flag.

File: src/js/tabs/firmware_flasher.js
Comment (haslinghuis) [reply src/js/tabs/firmware_flasher.js:836]: @limonspb DFU is not translated.

**Review by Asizon — APPROVED** [review]

**Review by Asizon — CHANGES_REQUESTED** [review]
Sorry but seems that reused flashing and hidden buttons issue is not yet adressed. 
To repoduce:
-Go to flasher tab
-With no coneccted board press on flash and configurator breaks.

I have attached working buikd screenshot to you can see how it should work.

image

**Comment by Asizon** [comment]
@hydra any test here with your concerned issues in https://github.com/betaflight/betaflight-configurator/issues/2526 ?

**Comment by blckmn** [comment]
AUTOMERGE: (FAIL)
 - github identifies PR as mergeable -> FAIL
 - assigned to a milestone -> PASS
 - cooling off period lapsed -> PASS
 - commit count less or equal to three -> PASS
 - `Don't merge` label NOT found -> PASS
 - at least one `RN:` label found -> PASS
 - `Tested` label found -> FAIL
 - assigned to an approver -> PASS
 - approver count at least three -> FAIL

**Comment by haslinghuis** [comment]
Thanks for testing @Asizon and pointing out how to reproduce. Removing the check in serial.js has solved it.

**Review by Asizon — DISMISSED** [review]

**Comment by Asizon** [comment]
Looks really good now @haslinghuis , how about change flashing progress text  "Initiated reboot to botloader..." when serial com failed?and go to previous state, not related to this PR but it would be a good fix.

**Comment by haslinghuis** [comment]
@Asizon have added tab refresh on serial port fail and added firmware version check on board verification.

**Review by Asizon — DISMISSED** [review]

**Comment by hydra** [comment]
> @hydra any test here with your concernings in #2526 ?

I have not had time to test this PR.

I would suggest that the PR author (@haslinghuis ) replicates the steps-to-reproduce in #2526 against this PR.

A new unknown target can be produced by simply creating a variant of an existing target or temporarily changing the name of an existing target.

The firmware build system is designed such that anyone can create new targets, making them official is another thing entirely, so it's an edge case that this PR should handle for sure.

**Comment by haslinghuis** [comment]
@hydra could you provide the steps to produce a variant or how to rename a target. Would love to test but I don't know how to create custom firmware apart from just `make TARGET=xxx` or `make unified_zip` on a specific branch.

**Comment by haslinghuis** [comment]
@hydra have changed the name in a config file and the last commit should resolve your issue.

**Comment by haslinghuis** [comment]
Not sure why the Windows test fails. It worked locally

---
image

File: src/js/tabs/firmware_flasher.js
Comment (mikeller) [reply src/js/tabs/firmware_flasher.js:866]: 'DFU' is a term STM uses in their documentation, so I doubt this should be translated.

**Comment by Asizon** [comment]
@haslinghuis new detect button width doenst match selectors width and I think it breaks style, Mcgiver idea is to put this button with small width between the selector and the i18n text. But the big button design is ok to me.

**Comment by haslinghuis** [comment]
@McGiverGim I came up with something like:

Screenshot from 2021-07-27 00-28-08

Should I make one tooltip instead?

**Comment by haslinghuis** [comment]
@McGiverGim I think you mean something like this:

Screenshot from 2021-07-27 01-09-08

**Comment by McGiverGim** [comment]
Sorry @haslinghuis I edited your message by error, I think I've recovered it ok.

Yes, this is one option. I've doubts between different options but this one is ok to me. Thanks for looking at it.

File: src/js/tabs/firmware_flasher.js
Comment (McGiverGim) [reply src/js/tabs/firmware_flasher.js:866]: I don't know if some alphabets, like Cyrillic, Chinese, Japanese, etc. can have some different symbols for the same.

**Review by McGiverGim — DISMISSED** [review]

**Comment by McGiverGim** [comment]
It seems the text at the right of the button is a little higher than the other texts.

**Comment by haslinghuis** [comment]
@McGiverGim fixed (including the tooltip):

Screenshot from 2021-07-28 00-01-59

**Review by McGiverGim — APPROVED** [review]

**Review by Asizon — APPROVED** [review]

**Comment by blckmn** [comment]
AUTOMERGE: (FAIL)
 - github identifies PR as mergeable -> FAIL
 - assigned to a milestone -> PASS
 - cooling off period lapsed -> PASS
 - commit count less or equal to three -> PASS
 - `Don't merge` label NOT found -> PASS
 - at least one `RN:` label found -> PASS
 - `Tested` label found -> PASS
 - assigned to an approver -> PASS
 - approver count at least three -> FAIL

**Review by limonspb — APPROVED** [review]

**Comment by blckmn** [comment]
AUTOMERGE: (PASS)
 - github identifies PR as mergeable -> PASS
 - assigned to a milestone -> PASS
 - cooling off period lapsed -> PASS
 - commit count less or equal to three -> PASS
 - `Don't merge` label NOT found -> PASS
 - at least one `RN:` label found -> PASS
 - `Tested` label found -> PASS
 - assigned to an approver -> PASS
 - approver count at least three -> PASS

**Pull request merged**
